airflow/dags/rxclass/dag.py

```diff
@@ -1,20 +1,192 @@
-from airflow_operator import create_dag
+import time
+import json
+import logging
+import threading
+from functools import partial
+from concurrent.futures import ThreadPoolExecutor
+from urllib.request import urlopen
+from urllib.error import HTTPError
 
-from rxclass.dag_tasks import get_rxcuis, extract_atc
-from common_dag_tasks import  get_ds_folder, get_data_folder, transform
+import pendulum
+import pandas as pd
+from pandas import json_normalize
 
-dag_id = 'rxclass_atc_to_product'
+from airflow.decorators import dag, task
+from airflow.operators.empty import EmptyOperator
 
-dag = create_dag(
+from sagerx import load_df_to_pg
+
+# Set up logging
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+
+class RateLimiter:
+    """
+    Restrict calls to a maximum of `max_calls` within `period` seconds.
+    Ensures we don't exceed API rate limits.
+    """
+    def __init__(self, max_calls, period):
+        self.max_calls = max_calls
+        self.period = period
+        self.calls = []
+        self.lock = threading.Lock()
+
+    def __call__(self, f):
+        def wrapped(*args, **kwargs):
+            with self.lock:
+                now = time.time()
+                # Remove calls that fell outside the time window
+                self.calls = [c for c in self.calls if now - c < self.period]
+                # If we've reached the limit, sleep until we're allowed to call again
+                if len(self.calls) >= self.max_calls:
+                    sleep_time = self.period - (now - self.calls[0])
+                    logging.info(f"Rate limit reached. Sleeping for {sleep_time:.2f} seconds.")
+                    time.sleep(sleep_time)
+                self.calls.append(time.time())
+            return f(*args, **kwargs)
+        return wrapped
+
+
+@RateLimiter(max_calls=20, period=1)  # Limit to 20 calls/second
+def fetch_json(url):
+    """
+    Fetch JSON from a URL, respecting the rate limit.
+    """
+    with urlopen(url) as response:
+        return json.loads(response.read())
+
+
+def process_concept(class_base_url, concept, max_retries=3, initial_delay=1):
+    """
+    For a given concept dict that includes 'rxcui', call the rxclass API
+    to retrieve class info. Implements retry logic to handle HTTP errors.
+    """
+    url = class_base_url + concept['rxcui']
+    for attempt in range(max_retries):
+        try:
+            cur_json = fetch_json(url)
+            class_data = cur_json['rxclassDrugInfoList']['rxclassDrugInfo']
+            # Return a list of merged data dicts
+            return [
+                dict(
+                    concept,
```

Comment:
One downside of concatenating the url upstream of this function is not being able to do things like this... but I think that's worth it for reusability.

Reply:
By "things like this" I mean returning the RXCUI from the query as part of the response.